### ISSUE TITLE
Truffle Decoder's second parameter of dependent contracts should be optional

### ISSUE THREAD

**Issue by adrianmcli** [trufflesuite/truffle#2820]
Calling this:

```js
const decoder = await Decoder.forContractInstance(truffleContractInstance);
```

Produces a cryptic error:

```js
TypeError: Cannot read property 'includes' of undefined
    at index.js:374
    at step (index.js:261)
    at Object.next (index.js:192)
    at index.js:164
...
```

It is only when you pass in an empty array as the second parameter does the error go away:

```js
const decoder = await Decoder.forContractInstance(truffleContractInstance, []);
```

**Comment by haltman-at** [comment]
So, is the behavior you intend that it should work the same as if an empty array was passed?  Hm.

Let me tell you what's planned for this right now.  The plan is that that final argument should be able to be a number of things allowing for all relevant contracts to be found -- not just a straight list of them, but also say a resolver or config object that's able to find them.  But that relies on some functionality that isn't implemented yet.

Anyway, I guess leaving it out having it act the same as `[]` makes sense, but I do want to avoid accidentally introducing some sort of inconsistency there.

**Comment by adrianmcli** [comment]
Forgive me for a second here as this is a bit subjective and a bit of a tangential rant, but I've never been a fan of APIs that have a bunch of non-[named parameters](https://en.wikipedia.org/wiki/Named_parameter). Having to know what the 1st, 2nd, 3rd argument is causes confusion and makes things a living hell whenever the API changes.

I'm not sure if it's too late, but I much prefer an API that allows you to simulate named parameters with an object (as it is much more clear and explicit):

```js
const decoder = await Decoder.forContractInstance({
  instance: truffleContractInstance,
  dependentArtifacts: [],
  resolver: myResolverFn
});
```

That being said, I'd be happy to accept a simple change to have an empty array as default:

```ts
export async function forContractInstance(
  contract: ContractInstanceObject,
  artifacts: Artifact[] = []
): Promise<ContractInstanceDecoder> {
  return await forContractAt(contract.constructor, contract.address, artifacts);
}
```

Basically just changing `artifacts: Artifact[]` to `artifacts: Artifact[] = []`, it should be a very simple fix.

**Comment by haltman-at** [comment]
Good news, my current plan is actually to handle the last argument by having it be an object specifying what it is!  (Not going to redo the other arguments though, that'd be a breaking change...)  With the existing form left in for compatibility.  So, yeah, I guess having it default to `[]` is a change I could make as well...

**Comment by adrianmcli** [comment]
That is good news!

**Comment by haltman-at** [comment]
So, looking at the two issues you submitted, the conclusion I'm coming to is that I didn't really optimize this for convenience in this just-a-simple-contract case, where there are no external structs or enums or contract types you need to deal with, no inheritance, you're not decoding events of unknown provenance, etc.  I was really focused on the opposite case (which I realize is *not* exactly convenient at the moment, but I was planning to make it so eventually).  But, I guess I can largely address that by the means you suggest, of making that argument optional and treating it as `[]`.  So, whew, problem avoided there.

**Comment by adrianmcli** [comment]
Yay, thanks @haltman-at !

**Comment by haltman-at** [comment]
Oh, huh, why is this still open?  Closing.
